### ISSUE TITLE
Type guards are not 'reset' after awaiting a promise

### ISSUE THREAD

**Issue by remag9330** [Microsoft/TypeScript#27900]
**TypeScript Version:**  3.2.0-dev.20181011 (though I found it while using 3.1)

**Search Terms:** async await type check change after await

**Code**

```ts
let elem: number | null = 5;

async function main() {
    const p1 = test()
    const p2 = breaker()
    await Promise.all([p1, p2])
}

async function test() {
    if (typeof elem === "number") {
        // Type checks elem is number, type is number.
        console.log(elem.toString())
        await delay(500)
        // Type of elem is still number, even though it could be set to null (in breaker) 
        console.log(elem.toString());
    }
}

function test2() {
    if (typeof elem === "number") {
        // Type of elem is number.
        console.log(elem.toString())
        delay(500).then(() => {
            // Type is now number | null, which is correct
            console.log(elem.toString())
        })
    }
}

async function breaker() {
    await delay(250)
    elem = null;
}

async function delay(ms: number) {
    return new Promise((resolve, reject) => setTimeout(resolve, ms))
}

main()
```

**Expected behavior:**
2 errors, one in test() after the await that elem is type number | null, and another in test2() in the callback of the delay promise that elem is type number | null

**Actual behavior:**
Only the second error is showing up. Commenting out test2 will compile fine but fail at runtime with "Cannot read property 'toString' of null" error.

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=let%20elem%3A%20number%20%7C%20null%20%3D%205%3B%0D%0A%0D%0Aasync%20function%20main()%20%7B%0D%0A%20%20%20%20const%20p1%20%3D%20test()%0D%0A%20%20%20%20const%20p2%20%3D%20breaker()%0D%0A%20%20%20%20await%20Promise.all(%5Bp1%2C%20p2%5D)%0D%0A%7D%0D%0A%0D%0Aasync%20function%20test()%20%7B%0D%0A%20%20%20%20if%20(typeof%20elem%20%3D%3D%3D%20%22number%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20Type%20checks%20elem%20is%20number%2C%20type%20is%20number.%0D%0A%20%20%20%20%20%20%20%20console.log(elem.toString())%0D%0A%20%20%20%20%20%20%20%20await%20delay(500)%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20Type%20of%20elem%20is%20still%20number%2C%20even%20though%20it%20could%20be%20set%20to%20null%20(in%20breaker)%20%0D%0A%20%20%20%20%20%20%20%20console.log(elem.toString())%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20test2()%20%7B%0D%0A%20%20%20%20if%20(typeof%20elem%20%3D%3D%3D%20%22number%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20Type%20of%20elem%20is%20number.%0D%0A%20%20%20%20%20%20%20%20console.log(elem.toString())%0D%0A%20%20%20%20%20%20%20%20delay(500).then(()%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20Type%20is%20now%20number%20%7C%20null%2C%20which%20is%20correct%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20console.log(elem.toString())%0D%0A%20%20%20%20%20%20%20%20%7D)%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Aasync%20function%20breaker()%20%7B%0D%0A%20%20%20%20await%20delay(250)%0D%0A%20%20%20%20elem%20%3D%20null%3B%0D%0A%7D%0D%0A%0D%0Aasync%20function%20delay(ms%3A%20number)%20%7B%0D%0A%20%20%20%20return%20new%20Promise((resolve%2C%20reject)%20%3D%3E%20setTimeout(resolve%2C%20ms))%0D%0A%7D%0D%0A%0D%0Amain())

**Related Issues:** 

In short, the issue seems to be that after an await, any type narrowing checks that occurred before are still in effect, which can cause problems if the variable changed to something else (in this case, set to null) after before the await finishes but after the check itself.

**Comment by nmain** [comment]
Isn't this just more or less the same thing as #9998?  Sure, it has `await` in it, but you can run into these sorts of problems in synchronous code as well; any function call could have all sorts of side effects in it.

**Comment by remag9330** [comment]
Thanks nmain, I can definitely see the similarities between these 2 issues. However, the biggest difference I can see between them is that in this example, there are 2 different ways to do (basically) the same thing (await and .then), but they produce different results from each other, and the inconsistency did cause an issue in our codebase recently.

Either way, I'm happy to let this be closed as a duplicate though, now knowing the other discussion exists.

**Comment by DanielRosenwasser** [comment]
#9998 is probably the best place to look at this discussion. In the general case, we try to "trust" that functions won't rip things out from under out feet and change something after it's been narrowed. I would say in that sense, the `await` example is exhibiting the "right" behavior given that assumption.

So why is the function expression case in `test2` exhibiting the same issue? Well, that's more of a symptom of not knowing whether a function is going to be executed immediately, after the next turn of the event loop, many times, once, or never. Because it's hard to reason about the lifetime of the variable we err on the conservative side. It's also consistent with the way we treat function declarations. It's someone inconsistent with the above goal, but it's ultimately the best balance to stay safe without having *too* many special cases. I could see us revisiting *this* behavior though.

**Comment by matthiasdailey-ccri** [comment]
Here's a similar example:
```ts
// This variable can be either undefined or an object.
let outsideVar: undefined | {a: {b: {c: any}}};

async function main () {
    // Set it to an object:
    outsideVar = {a: { b: { c: 123 }}};
    // Pause for 100ms. During this time, outsideVar will get set to undefined.
    await new Promise((resolve) => {
        setTimeout(resolve, 100);
    });
    // in the meantime, outsideVar gets unset but Typescript thinks it's still set.
    console.log(JSON.stringify(outsideVar));
    console.log(outsideVar.a.b.c); // No TypeScript error; there should be an error.
    // At runtime, outsideVar is undefined, and property access throws:
    // TypeError: Cannot read property 'a' of undefined
}

main();
outsideVar = undefined; // This will run before the `await` continues.

```

**Comment by thw0rted** [comment]
It looks like there hasn't been activity here for a while, is anybody still watching?  As far as I can tell, #9998 has a lot of discussion about flow control generally, but not much about `await` specifically.  I hope this is the right place to ask, since @DanielRosenwasser has said he "could see revisiting".

I came here to say that IMHO, code before and after an `await` statement is more akin to outer and inner function scopes, rather than before and after function calls that might have side effects.

````ts
declare target: string | undefined;
declare a: string[];
if (!target) { return; }
a.find(item => target.toLowerCase() === item.toLowerCase()); // Error, target could be undefined

getResourceFromServer().then(x => {
  console.log(target.toLowerCase() === x.toLowerCase());  // Error, target could be undefined
})

async function f() {
  if (!target) { return; }
  const x = await getResourceFromServer();
  target.toLowerCase() === x.toLowerCase(); // Totally fine?
}

setInterval(() => target = undefined, 100);
````

In the first and second cases (`Array#find` and `Promise#then`), it wouldn't be reasonable to expect Typescript to figure out when the anonymous functions will be invoked, so you have to be "pessimistic" as defined in #9998.  I would argue that the third example should work the same way, for the same reason -- you've ceded control of the event loop and you don't know when the rest of your function body is going to execute, so you have to assume that anything you previously knew about state is out the window now.

If it breaks too much to make such a big change in assumptions, maybe it's at least worth putting behind a flag?  Something in the `strict` family, perhaps?
